### ISSUE TITLE
model.fit() reports non-NaN float losses, but returns NaN when inferencing even though weights haven't changed

### ISSUE THREAD

**Issue by kongmunist** [tensorflow/tfjs#4149]
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): iOS ~13
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: yes, iPhone X
- TensorFlow.js installed from (npm or script link): script link
- TensorFlow.js version (use command below): 2.7.0
- Browser version: Safari 
- Tensorflow.js Converter Version: N/A

**Describe the current behavior**
I'm trying to fine-tune a complex computer vision neural net on the mobile browser, specifically iOS's Safari. The model architecture is really long and I don't think it's the problem, but basically it takes in two 128x128x3 images and an 8 vector, runs each input through some hidden layers, concats them and then runs them through some more hidden layers, and then outputs two numbers representing two normalized coordinates. 

My backend is webgl, and it's set before any other operations happen. I'm loading in the model (it was converted from a keras model using the python tfjs converter), then freezing every layer except the last few (everything here before the dense_3 has their trainable set to false): 
![Screen Shot 2020-10-28 at 12 04 37 AM](https://user-images.githubusercontent.com/29759597/97389833-baf79d80-18b1-11eb-981a-5a2a2a245325.png)
I then try to fit it using this code. It just compiles the model using sgd and runs fit, and at the end of every epoch evaluates it on random inputs and prints the current loss, :
```javascript
naturemodel.predict([tf.randomNormal([1,128,128,3]), tf.randomNormal([1,128,128,3]), tf.randomNormal([1,8])]).print()
var epochCount = 0
// Compile the model
naturemodel.compile({
  optimizer: tf.train.sgd(0.0000268),
  loss: 'meanSquaredError',
  metrics: ['mae', 'mse']
});
var n = 2;
console.log(tf.memory())

naturemodel.fit([tf.randomNormal([n,128,128,3]), tf.randomNormal([n,128,128,3]), tf.randomNormal([n,8])], [tf.randomNormal([n,2])], {
       epochs: 3,
       callbacks: {
          onEpochEnd: async (batch, logs) => {
                console.log('in batch predict ', naturemodel.predict([tf.randomNormal([1,128,128,3]),
                                                tf.randomNormal([1,128,128,3]), tf.randomNormal([1,8])]).arraySync())
                console.log(epochCount++, 'Loss: ' + logs.loss.toFixed(5));}
                }
     })

naturemodel.predict([tf.randomNormal([1,128,128,3]), tf.randomNormal([1,128,128,3]), tf.randomNormal([1,8])]).print()
```

So, originally, the losses, weights, and biases were all NaN after training no matter how small the learning rate was: 
![naturemodel.fit attempt with WEBGL_CPU_FORWARD left default, true](https://user-images.githubusercontent.com/29759597/97389970-1e81cb00-18b2-11eb-89d6-6c56c2486df1.png)

I then turned off the WEBGL_CPU_FORWARD flag with tf.ENV.set('WEBGL_CPU_FORWARD', false);. The model started reporting float losses (good!), but when evaluated on any inputs, the model returned NaN. This behavior continues even if I turn the WEBGL_CPU_FORWARD flag back to true. This is especially weird when you consider that the weights and biases of the last few layers were left unchanged (towards the bottom of the pic), so the model shouldn't have NaN output since it should be doing the same math as it was before. And also, if the loss is not NaN, why would the outputs be?
![naturemodel.fit attempt with WEBGL_CPU_FORWARD false](https://user-images.githubusercontent.com/29759597/97392697-cac6b000-18b8-11eb-9ab0-d63dacb0d3ce.png)

**Describe the expected behavior**
After running model.fit, I should to be able to use the model to do inference (especially if the weights are all still the same). I believe this has something to do with where the tensors are stored, or where the operations are occurring (cpu vs webgl), but I really don't know.

**Comment by kongmunist** [comment]
Have made some progress on a workaround. 

Duplicating the model after training doesn't fix the problem (making a new ```tf.model({inputs: naturemodel.inputs, outputs:naturemodel.outputs});```.

However, replacing the unfrozen hidden layers (everything in the architecture screenshow) with other dense layers does allow me to retrain and use the outputs. The only thing different is there's no batch norm layers in the new layers, however that could be a red herring.
```javascript
var modelcopy = tf.model({inputs: naturemodel.inputs, outputs: naturemodel.layers[33].output}); 

var tmplayer1 = tf.layers.dense({units: 32, activation: 'relu'});
var tmplayer2 = tf.layers.dense({units: 2});

var newoutput = tmplayer2.apply(tmplayer1.apply(naturemodel.layers[33].output))
var modelcopy = tf.model({inputs: naturemodel.inputs, outputs: newoutput}); // Model with new regression head attached
naturemodel = modelcopy;


```
